lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+      clang::MultiExprArg(), operatorNew, operatorDelete, /*Diagnose=*/false);
   if (findingAllocFuncFailed || !operatorNew || operatorNew->isDeleted() ||
       operatorNew->getAccess() == clang::AS_private ||
       operatorNew->getAccess() == clang::AS_protected)
-    return nullptr;
+    return synthesizedFactories;
 
   clang::QualType cxxRecordPtrTy = clangCtx.getPointerType(cxxRecordTy);
   // Adding `_Nonnull` to the return type of synthesized static factory
   bool nullabilityCannotBeAdded =
       clangSema.CheckImplicitNullabilityTypeSpecifier(
-          cxxRecordPtrTy, clang::NullabilityKind::NonNull,
-          cxxRecordDecl->getLocation(),
-          /*isParam=*/false,
-          /*OverrideExisting=*/true);
+          cxxRecordPtrTy, clang::NullabilityKind::NonNull, cxxRecordDeclLoc,
+          /*isParam=*/false, /*OverrideExisting=*/true);
   assert(!nullabilityCannotBeAdded &&
          "Failed to add _Nonnull specifier to synthesized "
          "static factory's return type");
 
   clang::IdentifierTable &clangIdents = clangCtx.Idents;
-  clang::IdentifierInfo *funcNameToSynthesize = &clangIdents.get(
-      ("__returns_" + cxxRecordDecl->getNameAsString()).c_str());
-  clang::FunctionProtoType::ExtProtoInfo EPI;
-  clang::QualType funcTypeToSynthesize =
-      clangCtx.getFunctionType(cxxRecordPtrTy, {}, EPI);
-
-  clang::CXXMethodDecl *synthesizedCxxMethodDecl = clang::CXXMethodDecl::Create(
-      clangCtx, const_cast<clang::CXXRecordDecl *>(cxxRecordDecl),
-      cxxRecordDecl->getLocation(),
-      clang::DeclarationNameInfo(funcNameToSynthesize,
-                                 cxxRecordDecl->getLocation()),
-      funcTypeToSynthesize,
-      clangCtx.getTrivialTypeSourceInfo(funcTypeToSynthesize), clang::SC_Static,
-      /*UsesFPIntrin=*/false, /*isInline=*/true,
-      clang::ConstexprSpecKind::Unspecified, cxxRecordDecl->getLocation());
-  assert(synthesizedCxxMethodDecl &&
-         "Unable to synthesize static factory for c++ foreign reference type");
-  synthesizedCxxMethodDecl->setAccess(clang::AccessSpecifier::AS_public);
-
-  if (!hasImmortalAttrs(cxxRecordDecl)) {
-    clang::SwiftAttrAttr *returnsRetainedAttrForSynthesizedCxxMethodDecl =
-        clang::SwiftAttrAttr::Create(clangCtx, "returns_retained");
+
+  for (clang::CXXConstructorDecl *selectedCtorDecl : cxxRecordDecl->ctors()) {
+    if (selectedCtorDecl->isDeleted() ||
+        selectedCtorDecl->getAccess() == clang::AS_private ||
+        selectedCtorDecl->getAccess() == clang::AS_protected ||
+        selectedCtorDecl->isCopyOrMoveConstructor())
+      continue;
+
+    unsigned int ctorParamCount = selectedCtorDecl->getNumParams();
+    std::string funcName = "__returns_" + cxxRecordDecl->getNameAsString();
+    if (ctorParamCount > 0)
+      funcName += "_" + std::to_string(ctorParamCount) + "_params";
+    clang::IdentifierInfo *funcNameToSynthesize = &clangIdents.get(funcName);
+
+    llvm::SmallVector<clang::QualType, 4> paramTypes;
```

Comment:
You might not need to create a new vector here. You could get the type of the ctor, which should be a `clang::FunctionProtoType`. And you could call `getParamTypes` which should return an `ArrayRef` to already allocated storage that has this information. This can shave off some allocations making the code more efficient (especially with ctors that have lots of parameters).